File: app/src/main/java/com/nikola/jakshic/dagger/viewModel/PeerViewModel.java
package com.nikola.jakshic.dagger.viewModel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.paging.PagedList;

import com.nikola.jakshic.dagger.Status;
import com.nikola.jakshic.dagger.model.Peer;
import com.nikola.jakshic.dagger.repository.PeerRepository;

import javax.inject.Inject;

public class PeerViewModel extends ViewModel {

    private LiveData<PagedList<Peer>> list;
    private PeerRepository repository;
    private MutableLiveData<Status> status;
    private boolean initialFetch;

    @Inject
    public PeerViewModel(PeerRepository repository) {
        this.repository = repository;
        status = new MutableLiveData<>();
        status.setValue(Status.LOADING);
    }

    public void initialFetch(long id) {
        if (!initialFetch) {
            list = repository.fetchByGames(id);
            repository.fetchPeers(status, id);
            initialFetch = true;
        }
    }

    // Get list from DB sorted by games
    public void sortByGames(long id) {
        list = repository.fetchByGames(id);
    }

    // Get list from DB sorted by win rate
    public void sortByWinrate(long id) {
       list = repository.fetchByWinrate(id);
    }

    // Fetch peers from network and store them into DB
    public void fetchPeers(long id) {
        repository.fetchPeers(status, id);
    }

    // Expose peer data from DB
    public LiveData<PagedList<Peer>> getPeers() {
        return list;
    }

    // Expose network status
    public MutableLiveData<Status> getStatus() {
        return status;
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/profile/peers/PeerViewModel.kt
package com.nikola.jakshic.dagger.ui.profile.peers

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.nikola.jakshic.dagger.repository.PeerRepository
import com.nikola.jakshic.dagger.ui.ScopedViewModel
import com.nikola.jakshic.dagger.ui.Status
import com.nikola.jakshic.dagger.vo.Peer
import kotlinx.coroutines.launch
import javax.inject.Inject

class PeerViewModel @Inject constructor(private val repository: PeerRepository) : ScopedViewModel() {

    lateinit var list: LiveData<List<Peer>>
        private set

    private val _status = MutableLiveData<Status>()
    val status: LiveData<Status>
        get() = _status

    private var initialFetch = false

    private val onSuccess: () -> Unit = { _status.value = Status.SUCCESS }
    private val onError: () -> Unit = { _status.value = Status.ERROR }

    fun initialFetch(id: Long) {
        if (!initialFetch) {
            initialFetch = true
            fetchPeers(id)
            list = repository.getPeersLiveDataByGames(id)
        }
    }

    fun fetchPeers(id: Long) {
        launch {
            _status.value = Status.LOADING
            repository.fetchPeers(id, onSuccess, onError)
        }
    }

    fun sortByGames(id: Long) {
        list = repository.getPeersLiveDataByGames(id)
    }

    fun sortByWinRate(id: Long) {
        list = repository.getPeersLiveDataByWinrate(id)
    }
}